src/System.Private.CoreLib/shared/System/BitOps.cs

```diff
@@ -8,31 +8,62 @@
+        // https://github.com/dotnet/roslyn/pull/24621
+
+        private static ReadOnlySpan<byte> s_TrailingZeroCountDeBruijn => new byte[32]
+        {
+            00, 01, 28, 02, 29, 14, 24, 03,
+            30, 22, 20, 15, 25, 17, 04, 08,
+            31, 27, 13, 23, 21, 19, 16, 07,
+            26, 12, 18, 06, 11, 05, 10, 09
+        };
+
+        /// <summary>
+        /// Count the number of trailing zero bits in an integer value.
+        /// Similar in behavior to the x86 instruction TZCNT.
+        /// </summary>
+        /// <param name="value">The value.</param>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static int TrailingZeroCount(int matches)
+        public static uint TrailingZeroCount(int value)
+            => TrailingZeroCount((uint)value);
+
+        /// <summary>
+        /// Count the number of trailing zero bits in an integer value.
+        /// Similar in behavior to the x86 instruction TZCNT.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint TrailingZeroCount(uint value)
         {
             if (Bmi1.IsSupported)
             {
-                return (int)Bmi1.TrailingZeroCount((uint)matches);
-            }
-            else // Software fallback
-            {
-                // https://graphics.stanford.edu/~seander/bithacks.html#ZerosOnRightMultLookup
-                // uint.MaxValue >> 27 is always in range [0 - 31] so we use Unsafe.AddByteOffset to avoid bounds check
-                return Unsafe.AddByteOffset(
-                    ref MemoryMarshal.GetReference(TrailingCountMultiplyDeBruijn),
-                    ((uint)((matches & -matches) * 0x077CB531U)) >> 27);
+                // Note that TZCNT contract specifies 0->32
+                return Bmi1.TrailingZeroCount(value);
             }
-        }
 
-        private static ReadOnlySpan<byte> TrailingCountMultiplyDeBruijn => new byte[32]
-        {
-            0, 1, 28, 2, 29, 14, 24, 3, 30, 22, 20, 15, 25, 17, 4, 8,
-            31, 27, 13, 23, 21, 19, 16, 7, 26, 12, 18, 6, 11, 5, 10, 9
-        };
+            // Using deBruijn sequence, k=2, n=5 (2^5=32)
+            const uint deBruijn = 0b_0000_0111_0111_1100_1011_0101_0011_0001u; // 0x077CB531u
+            uint ix = (uint)((value & -value) * deBruijn) >> 27;
+
+            // uint.MaxValue >> 27 is always in range [0 - 31] so we use Unsafe.AddByteOffset to avoid bounds check
+            ref byte tz = ref MemoryMarshal.GetReference(s_TrailingZeroCountDeBruijn);
+            uint count = Unsafe.AddByteOffset(ref tz, (IntPtr)ix);
+
+            // Above code has behavior 0->0, so special-case in order to match intrinsic path
+
+            // Branchless equivalent of: c32 = value == 0 ? 32 : 0
+            bool is0 = value == 0;
+            uint c32 = Unsafe.As<bool, byte>(ref is0) * 32u; // 0|1 x 32
```

Comment:
This has a stack spill. Is it really an improvement over a branch?

I do not think it is a good idea to try to do manual micro-optimizations like this. This should just use regular `? :` with assumption that the JIT will start emitting conditional mov one day.

Reply:
This kind of optimizations aren't even guaranteed to be optimizations. I've seen enough cases where attempting the eliminate a branch just makes the code slower. It's highly dependent on the input distribution.

Reply:
> With the usual caveats for microbenchmarks, I tested the perf when I implemented it

How does the test look like? What are the input values?

Reply:
```csharp
        private const int _iterations = 5000;
        private const int N = ushort.MaxValue;

        // Try prevent folding by using volatile non-readonly non-constant
        private static volatile bool s_true = true;
        private static volatile bool s_false = false;

        [Benchmark(Baseline = true, OperationsPerInvoke = _iterations * N)]
        public static ulong Branch()
        {
            ulong sum = 0;

            for (int i = 0; i < _iterations; i++)
            {
                for (int n = 0; n <= N; n++)
                {
                    // Used various distributions for condition below, eg n % 3 <= 1, n % 5 >= 2
                    bool condition = n % 2 == 0 ? s_true : s_false;
                    // Other method uses:
                    // sum += System.Runtime.CompilerServices.Unsafe.As<bool, byte>(ref condition) * 32u;
                    sum += condition ? 32u : 0u;
                    sum++;
                }
            }

            return sum;
        }
```

Reply:
You should measure the performance of the whole `TrailingZeroCount` function with each option.

There is a lot of interactions at this level between different parts of the function, both in the optimizations as well as in the hardware. It is not meaningful to measure this construct without the context that it is used in.

Reply:
Good call @jkotas. Measuring the whole function had a significant perf regression; 1.4x slower. 
I will revert to using an `if`. 
```
     Method |     Mean |     Error |    StdDev | Ratio | RatioSD |
----------- |---------:|----------:|----------:|------:|--------:|
      Guard | 2.297 ns | 0.0338 ns | 0.0316 ns |  1.00 |    0.00 | -- guard clause
     Branch | 2.844 ns | 0.0494 ns | 0.0462 ns |  1.24 |    0.03 | -- ? : after calc
 UnsafeImpl | 3.279 ns | 0.0444 ns | 0.0347 ns |  1.43 |    0.02 | -- branchless
```